semgrep-core/matching/Matching_generic.ml

```diff
@@ -218,16 +218,22 @@ let equal_ast_binded_code (a: AST.any) (b: AST.any) : bool =
        *)
       let a = Lib.abstract_position_info_any a in
       let b = Lib.abstract_position_info_any b in
-      let res = a =*= b in
-      if !Flag.debug_matching && not res
-      then begin
-        pr2 (spf "A = %s" (str_of_any a));
-        pr2 (spf "B = %s" (str_of_any b));
-      end;
-      res
-
+      a =*= b
+  | A.I _, A.E (A.Id (b_id, _)) ->
```

Comment:
@aryx @mjambon I'm open to suggestions on how to do this better :)

Reply:
We might need also to handle the symetric pattern
A.E (A.ID (a_id, _)), A.I _

Reply:
Maybe add a comment such as
"you should prefer to add metavar as expression (A.E), not id (A.I), see Generic_vs_generic.m_ident_and_id_info_add_in_env_Expr  but in some cases you have no choice
and you need to match an expression metavar with an id metavar.
For example, we want the pattern 'const $X = foo.$X'  to match 'const bar = foo.bar' (this is useful in the Javascript transpilation context of complex pattern parameter)."